main/test/flix/Test.Exp.Jvm.NewObject.flix

```diff
@@ -52,22 +42,87 @@ namespace Test/Exp/Jvm/NewObject {
+    import flix.test.TestDefaultMethods.methodWithDefaultImplementation(Int32): Int32;
+    let anon = object ##flix.test.TestDefaultMethods {
+      def methodWithNoImplementation(_this: ##flix.test.TestDefaultMethods, x: Int32): Int32 = x + 1
+    };
+    methodWithNoImplementation(anon, 1) == 2 and methodWithDefaultImplementation(anon, 1) == 43
+
+  @test
+  def testDefaultMethods02(): Bool & Impure =
+    import flix.test.TestDefaultMethods.methodWithNoImplementation(Int32): Int32;
+    import flix.test.TestDefaultMethods.methodWithDefaultImplementation(Int32): Int32;
+    let anon = object ##flix.test.TestDefaultMethods {
+      def methodWithNoImplementation(_this: ##flix.test.TestDefaultMethods, x: Int32): Int32 = x + 1
+      def methodWithDefaultImplementation(_this: ##flix.test.TestDefaultMethods, x: Int32): Int32 = x + 2
+    };
+    methodWithNoImplementation(anon, 1) == 2 and methodWithDefaultImplementation(anon, 1) == 3
+
+  @test
+  def testGenerics01(): Bool & Impure =
+    import static flix.test.TestGenerics.runTest(##flix.test.TestGenerics): Bool;
+    import java.lang.Integer.toString(): String;
+    let anon = object ##flix.test.TestGenerics {
+      def methodOne(_this: ##flix.test.TestGenerics, x: ##java.lang.Object): ##java.lang.Object = 
+        let str = x as String;
+        "${str}, ${str}" as ##java.lang.Object
+      def methodTwo(_this: ##flix.test.TestGenerics, x: ##java.lang.Object, y: ##java.lang.Object): ##java.lang.Object =
+        let str = x as String;
+        let int = y as ##java.lang.Integer;
+        "${str}, ${toString(int)}" as ##java.lang.Object
+    };
+    runTest(anon)
+
+  @test
+  def testClassWithDefaultConstructor01(): Bool & Impure =
+    import flix.test.TestClassWithDefaultConstructor.abstractMethod(Int32): Int32;
+    import flix.test.TestClassWithDefaultConstructor.concreteMethod(String): String;
+    let anon = object ##flix.test.TestClassWithDefaultConstructor {
+      def abstractMethod(_this: ##flix.test.TestClassWithDefaultConstructor, x: Int32): Int32 = x + 1
+    };
+    abstractMethod(anon, 1) == 2 and
+      concreteMethod(anon, "bar") == "foobar"
+
   @test
-  def testImplementInterface08(): Bool & Impure =
-    import java.lang.Comparable.compareTo(##java.lang.Object): Int32 & Pure;
-    let anon = implementComparable();
-    compareTo(anon, anon as ##java.lang.Object) == 0
+  def testClassWithDefaultConstructor02(): Bool & Impure =
+    import flix.test.TestClassWithDefaultConstructor.abstractMethod(Int32): Int32;
+    import flix.test.TestClassWithDefaultConstructor.concreteMethod(String): String;
+    let anon = object ##flix.test.TestClassWithDefaultConstructor {
+      def abstractMethod(_this: ##flix.test.TestClassWithDefaultConstructor, x: Int32): Int32 = x + 1
+      def concreteMethod(_this: ##flix.test.TestClassWithDefaultConstructor, y: String): String = "flix: ${y}"
+    };
+    abstractMethod(anon, 1) == 2 and
+      concreteMethod(anon, "bar") == "flix: bar"
```

Comment:
As a general comment, I worry whether these tests are too complicated.

See for example:

https://github.com/flix/flix/blob/master/main/test/flix/Test.Exp.Jvm.PutStaticField.flix

(Of course those tests can get away with being simpler, but it might still be nice to split some tests up into "smaller" tests). 

Also, I am worried about the use of `runTest`. Don't get me wrong, such helper functions are usually fine in normal test cases, but when writing a compiler, and a thing goes wrong, it can become extremely messy if helper functions are used.

That's why most other tests try to use as few constructs as humanly possible. 

For example, I would be more inclined towards e.g.:

```
public interface TestInt32Interface {
  int get();
  void set(int x);
}
```

and then a batch of test cases that use just this interface.

Repeat 7 (8?) x for all types.

I know this is very tedious to write-- (possibly we could ask a student programmer)-- but it is much better for when things go wrong. Otherwise what can happen is that **all** test cases fail when we make just a small breakage, for say Int32, or whatever.

Reply:
We don't have to address this specific comment before I merge this PR, as work in progress.

Reply:
I have no objection to splitting this into multiple smaller tests (understand the motivation for doing so, and no need to palm it off on a student!).

Reply:
The motivation behind `runTest` was that I wanted to test "from" Java. I worry that tests written in Flix could easily depend upon vagaries of Flix?

Reply:
Yes, but you depend on strings, and on the `ToString` type class which means suddenly you touch a lot of the language inadvertently :)

Reply:
Right. But that's deliberate. It was only by doing this that I uncovered the bug that you fixed in #4342. We need to have *some* test that tests this kind of thing?

Reply:
But could we not call each method one at a time and check that it returns 42 (if it is an int) and so forth?

Reply:
> But could we not call each method one at a time and check that it returns 42 (if it is an int) and so forth?

Yes, of course we could.

There are two reasons why I'm reticent to do so:

Firstly, calling Flix from Flix feels like cheating. What we're implementing here is something that's intended to support Java interop. Just because Flix can call Flix doesn't necessarily mean that Java can call Flix?

Secondly, avoiding the use of `ToString` or similar would mean that I wouldn't have found the bug that you fixed in #4342.

Nevertheless, if this is what you would prefer, I'll defer to your judgement.

Reply:
(unless you're talking about splitting it out into multiple separate tests, in which case I've already said that I agree?)

Reply:
Oh- maybe we are talking past each other then.

Reply:
OK, so just to make sure that we're on the same page:

I'll split the current tests into multiple separate tests, but still retain the Java `runTest` method?

Reply:
(and still retain the use of `ToString` in the test case that tests multiple method parameters—I want to retain this because without it, there's no test to ensure that we're calculating parameter stack offsets correctly)

Reply:
OK, I miss-read what runTest does. We can have something like that, but I would have one for each interface.

Reply:
I would just like to have tests that are very incremental:

- Implement a method and check its return value.
- Implement a method and pass an argument to it.
- Implement a method where I use captured variables, type classes, etc.

Reply:
Gotcha 👍

I'll see about breaking things down like this tomorrow.